pantheon-bundle/src/main/java/com/redhat/pantheon/servlet/module/ModuleVersionUpload.java

```diff
@@ -106,14 +105,29 @@ protected void doRun(SlingHttpServletRequest request, PostResponse response, Lis
                     .draft().getOrCreate();
 
             // Check if the content is the same as what is hashed already
-            HashCode incomingSrcHash = hash(asciidocContent);
+            HashCode incomingSrcHash =
+                ServletUtils.handleParamAsStream(request, "asciidoc",
+                        inputStream -> {
+                            try {
+                                return JcrResources.hash(inputStream);
+                            } catch (IOException e) {
+                                throw new RuntimeException(e);
+                            }
+                        });
             String storedSrcHash = draftSrc.hash().get();
             // If the source content is the same, don't update it
             if(incomingSrcHash.toString().equals( storedSrcHash )) {
                 responseCode = HttpServletResponse.SC_NOT_MODIFIED;
             } else {
-                draftSrc.jcrContent().getOrCreate()
-                        .jcrData().set(asciidocContent);
+                ServletUtils.handleParamAsStream(request, "asciidoc", encoding,
+                        inputStream -> {
+                            Session session = resolver.adaptTo(Session.class);
+                            draftSrc.jcrContent().getOrCreate()
+                                    .jcrData().toFieldType(InputStream.class)
+                                    .set(inputStream);
```

Comment:
Isn't this fancy! What's going on here? Is this a magic way to convert the Stream to a String so that it can be persisted in the jcr?

Also, are you sure this works? My concern is that the Stream has already been consumed by the hashing method, and nothing ever reset it?

Reply:
It's not that magical (you can see that it's just calling some JCR utilities), but it is exactly for the purpose of what you described.

It does work (at least tested locally), but now that you ask the question, the input stream is not being closed anywhere so perhaps there is still a memory leak here.

Reply:
reviewed again... there are no problems because the input stream is being passed as the argument to the function, and it's being closed in the method after being used up by the function. All good 😊

Reply:
Can you point me to "the function" that you are referring to? I am genuinely curious how this works. AFAIK, Sling has to resolve this stream to some base type in order to be able to persist it, but how does it do that? It has no hint in this code that the stream here represents a string. It could be a stream of bytes, or anything, right? Did *we* write the method that resolves this and *we* are making that assumption? Or is something else going on?

Reply:
You are correct, the input stream can be of anything, and because we are 'setting' an InputStream into a jcr field this results in jackrabbit slurping the contents of the stream into the node's field. 
Now we (as in the caller) determines what stream is being passed to the 'set' method, in this case it's a character stream (UTF-8 or whatever encoding we decided on) directly from the sling request. 
Because the content is being pushed to the sling API as an `nt:file` sling smartly creates a temporary file where it stores the content, and in practice this is the stream that we are getting.